Add entries for third party plugin dev portal (#21)
Ableton/ableton.github.io: third_party_plugin_dev_portal/VST3presetSwitchingPerformEditAlternatives.md

---
layout: link
active_page: VST3presetSwitchingPerformEditAlternatives.
title: Recommendations to VST3 plugins manufacturers regarding bulk calls to performEdit
---

# Recommendations to VST3 plugins manufacturers regarding bulk calls to performEdit
## Introduction

For some VST3 plugins, Ableton receives reports of user interface lag when users switch presets from the plugin editor. After investigation, it seems that some VST3 manufacturers implement preset changes from the VST3 editor by calling performEdit for all parameters in order to notify both the DAW and the VST3 controller counterpart of its new state.

> Live needs time to react to performEdit in order for example to provide shadow-mapping of unmapped parameters, undo-redo.

Reviewer: It's not so clear what "shadow mapping of unmapped parameters" means. Maybe put "for our parameter mapping system"?

Author: I agree that "shadow mapping of unmapped parameters" is not a VST3 feature, nor a widely known feature of Live.

```suggestion
Live needs time to react to performEdit in order for our parameter mapping system.
```

Unfortunately, we receive calls from performEdit synchronously, and cannot optimize for situations where many successives calls of performEdit will happen.

## Steinberg’s recommendations: editor uses private communication to send new values to processor

> This solution is about using private communication (for example sending custom IMessage) from editor to component with all the new values to set before calling restartComponent and making sure that the controller will return the correct information before making a call to restartComponent(kParamValuesChanged).

Reviewer: sending a custom IMessage

Author: I don't understand this message. What change are you suggesting?

Reviewer: This solution is about using private communication (for example sending a custom IMessage) from editor to component with all the new values to set before calling restartComponent and making sure that the controller will return the correct information before making a call to restartComponent(kParamValuesChanged).

Reviewer: there is an "a" missing

Author: ```suggestion
This solution is about using private communication (for example sending a custom IMessage) from editor to component with all the new values to set before calling restartComponent and making sure that the controller will return the correct information before making a call to restartComponent(kParamValuesChanged).
```

> See below some quotes from Steinberg senior developer Arne Scheffler about it:

Reviewer: I don't think we should mention Arne personally like this

Author: how should that sentence look like ?

something like "See below some quotes from a Steinberg senior developer about it:" ?

### private communication (custom IMessage)

[https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#private-communication](https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#private-communication)

Data that is unknown to the host can be transmitted by means of messages. The communication interfaces are
* Steinberg::Vst::IConnectionPoint: The host establishes a connection between processor and controller.
* Steinberg::Vst::IMessage: Represents a message to send to the counterpart.
* Steinberg::Vst::IAttributeList: A list of attributes belonging to a message.

Please note that messages from the processor to the controller must not be sent during the process call, as this would not be fast enough and would break the real time processing. Such tasks should be handled in a separate timer thread.

> ## Implementation hints from open source frameworks

Reviewer: Not sure we should include this section at all to be honest.

Reviewer: because frameworks are constantly changing... they should all implement the VST3 API as correctly as possible so providing a good example with the VST3 SDK is enough imo.

Reviewer: > frameworks are constantly changing = more possibility of out of date info on our docs pages

### VST3SDK
#### open303

In this example, the processor sends the pattern to the editor. The other way around should be relatively similar.

[https://github.com/scheffle/open303/blob/850c98186db28829f9ff6c8254966cdb6bf334e6/Source/VST3/o303processor.cpp#L185-L203](https://github.com/scheffle/open303/blob/850c98186db28829f9ff6c8254966cdb6bf334e6/Source/VST3/o303processor.cpp#L185-L203)

```cpp
void sendPatternToController (int patternIndex)
{
	if (!peerConnection)
		return;

	const auto& pattern = open303Core.sequencer.getPattern (patternIndex);
	assert (pattern != nullptr);

	vst3utils::message msg (owned (allocateMessage ()));
	if (!msg.is_valid ())
		return;
	msg.set_id (msgIDPattern);
	auto attributes = msg.get_attributes ();
	if (!attributes.is_valid ())
		return;
	PatternData data = toPatternData (*pattern);
	attributes.set (msgIDPattern, data);
	peerConnection->notify (msg);
}
```

> #### IPlug2

Author: Paired with Oli: just remove the example for now. Maybe later some better example come.
